File: leetcode/python/Best Time to Buy and Sell Stock/Best_Time_to_Buy_and_Sell_Stock.py
class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        if not prices:
            return 0
        min_price = prices[0]
        max_profit = 0
        for i,p in enumerate(prices):
            max_profit = max(p - min_price,max_profit)
            min_price = min(p,min_price)
        return max_profit

File: leetcode/cpp/BestTimetoBuyandSellStock.cpp
//
// Created by zz on 2017/10/9.
// Time : O(n)
// Space : O(1)

#include <vector>
#include <limits>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int> &prices) {

        if (prices.empty()) {
            return 0;
        }


        int hold = numeric_limits<int>::min();
        int release = numeric_limits<int>::min();
        for (int price : prices) {
            hold = max(-price, hold);
            release = max(release, hold + price);
        }

        return release;
    }
};

